Source/DafnyRuntime/DafnyRuntime.go

```diff
@@ -692,183 +692,412 @@ func (seq Seq) VerbatimString(asLiteral bool) string {
 }
 
-// Equals compares two arrays for equality.  Values are compared using
-// dafny.AreEqual.
-func (array *Array) Equals(array2 *Array) bool {
-  if array == array2 {
-    return true
+/***** ArrayStruct is default implementation of the Array interface. *****/
+
+type ArrayStruct struct {
+  contents []interface{} // stored as a flat one-dimensional slice
+  dims     []int
+}
+
+func (_this ArrayStruct) dimensionCount() int {
+  return len(_this.dims)
+}
+
+func (_this ArrayStruct) dimensionLength(dim int) int {
+  return _this.dims[dim]
+}
+
+func (_this ArrayStruct) ArrayGet1(index int) interface{} {
+  return _this.contents[index]
+}
+
+func (_this ArrayStruct) ArraySet1(value interface{}, index int) {
+  _this.contents[index] = value
+}
+
+func (_this ArrayStruct) ArrayGet1Byte(index int) byte {
+  panic("cannot specialize general ArrayGet with byte")
+}
+
+func (_this ArrayStruct) ArraySet1Byte(value byte, index int) {
+  panic("cannot specialize general ArraySet with byte")
+}
+
+func (_this ArrayStruct) ArrayGet1Char(index int) Char {
+  panic("cannot specialize general ArrayGet with char")
+}
+
+func (_this ArrayStruct) ArraySet1Char(value Char, index int) {
+  panic("cannot specialize general ArraySet with char")
+}
+
+func (_this ArrayStruct) anySlice(lo, hi Int) []interface{} {
+  if lo.IsNilInt() && hi.IsNilInt() {
+    return _this.contents
   }
-  if array == nil || array2 == nil {
-    return false // we already know they're not equal as pointers
+  if lo.IsNilInt() {
+    return _this.contents[:hi.Int()]
   }
-  if len(array.dims) != len(array2.dims) {
-    return false
+  if hi.IsNilInt() {
+    return _this.contents[lo.Int():]
   }
-  for i, d := range array.dims {
-    if d != array2.dims[i] {
-      return false
-    }
+  return _this.contents[lo.Int():hi.Int()]
+}
+
+// ArrayStruct implements the EqualsGeneric interface.
+func (_this ArrayStruct) EqualsGeneric(other interface{}) bool {
+  otherArray, ok := other.(*ArrayStruct)
+  if !ok {
+    return false
   }
+  return &_this.dims[0] == &otherArray.dims[0]
+}
+
+/***** ArrayForByte is the Array interface specialized for byte. *****/
+
+type ArrayForByte struct {
+  contents []byte // stored as a flat one-dimensional slice
+  dims     []int
+}
 
-  return sliceIsPrefixAfterLengthCheck(array.contents, array2.contents)
+func (_this ArrayForByte) dimensionCount() int {
+  return len(_this.dims)
 }
 
-// EqualsGeneric implements the EqualsGeneric interface.
-func (array *Array) EqualsGeneric(other interface{}) bool {
-  array2, ok := other.(*Array)
-  return ok && array.Equals(array2)
+func (_this ArrayForByte) dimensionLength(dim int) int {
+  return _this.dims[dim]
 }
 
-func (array *Array) findIndex(ixs ...int) int {
-  i := 0
-  size := 1
-  for d := len(array.dims) - 1; d >= 0; d-- {
-    i += size * ixs[d]
-    size *= array.dims[d]
-  }
-  return i
+func (_this ArrayForByte) ArrayGet1(index int) interface{} {
+  return _this.contents[index]
```

Comment:
Shouldn't this also be a panic? Or do we need it to support type variables?

Reply:
I don't understand your comment. Are you proposing that `ArrayGet1` check to make sure the array has just one dimension and panic if it's not?

Reply:
I'm saying that if `ArrayStruct.ArrayGet1Byte` panics because it's not expected (even though it could conditional work by downcasting), I would expect that `ArrayForByte.ArrayGet1` should also panic since it's also a mismatch from expectations.

I then wondered if it's possible we need this because of type variables, because we might be holding on to an optimized byte array without knowing it. But then AFAICT `array<byte>` is not a subtype of `array<T>` in Dafny unless `T` is in fact `byte`, so I think a panic here would be safe and catch bugs.

Reply:
The reason is indeed because of type variables. Here's an example where we might be holding on to an optimized byte array without knowing it:

``` dafny
newtype byte = x | 0 <= x < 256

method Main() {
  var byteArray := new byte[100];
  var intArray := new int[100];

  var _ := byteArray[5]; // this calls ArrayGet1Byte on the optimize byte array
  byteArray[5] := 64; // similarly, this calls ArraySet1Byte
  var _ := intArray[5]; // this calls ArrayGet1
  intArray[5] := 64; // similarly, this calls ArraySet1

  var _ := UpdateArrayElement(byteArray, 5, 65);
  var _ := UpdateArrayElement(intArray, 5, 65);
}

method UpdateArrayElement<X>(arr: array<X>, j: nat, x: X) returns (previous: X)
  requires j < arr.Length
  modifies arr
{
  previous := arr[j]; // this always calls ArrayGet1, even on an optimized array
  arr[j] := x; // similarly, this always calls ArraySet1
}
```